gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/promotion/PromotionServiceImpl.java

```diff
@@ -237,42 +242,35 @@ public PromotionEntity createOrUpdate(PromotionEntity promotionEntity) {
     public PromotionEntity processPromotion(
         final String organizationId,
         final String environmentId,
-        String promotion,
+        String promotionId,
         boolean accepted,
         String user
     ) {
         try {
-            final Promotion existing = promotionRepository.findById(promotion).orElseThrow(() -> new PromotionNotFoundException(promotion));
+            final Promotion promotion = promotionRepository
+                .findById(promotionId)
+                .orElseThrow(() -> new PromotionNotFoundException(promotionId));
 
-            EnvironmentEntity environment = environmentService.findByCockpitId(existing.getTargetEnvCockpitId());
+            EnvironmentEntity environment = environmentService.findByCockpitId(promotion.getTargetEnvCockpitId());
 
-            existing.setStatus(accepted ? PromotionStatus.ACCEPTED : PromotionStatus.REJECTED);
+            promotion.setStatus(accepted ? PromotionStatus.ACCEPTED : PromotionStatus.REJECTED);
 
-            final PromotionQuery promotionQuery = new PromotionQuery();
-            promotionQuery.setStatuses(Collections.singletonList(PromotionEntityStatus.ACCEPTED));
-            promotionQuery.setTargetEnvCockpitIds(singletonList(existing.getTargetEnvCockpitId()));
-            promotionQuery.setTargetApiExists(true);
-            promotionQuery.setApiId(existing.getApiId());
+            JsonNode apiDefinition = objectMapper.readTree(promotion.getApiDefinition());
 
-            List<PromotionEntity> previousPromotions = search(promotionQuery, new SortableImpl("created_at", false), null).getContent();
+            ApiEntity apiToUpdate = findAlreadyPromotedTargetApi(environment, promotion, apiDefinition);
 
-            // Should create a new API if there is no previous promotion for this API or if the API existed once (after a promotion) but has been deleted since
-            boolean shouldCreate =
-                CollectionUtils.isEmpty(previousPromotions) || !apiService.exists(previousPromotions.get(0).getTargetApiId());
-
-            if (PromotionStatus.ACCEPTED.equals(existing.getStatus())) {
+            if (PromotionStatus.ACCEPTED.equals(promotion.getStatus())) {
                 ApiEntity promoted = null;
 
                 // FIXME: All the methods should take then env id as input instead of relying on GraviteeContext.getCurrentEnv
                 GraviteeContext.setCurrentEnvironment(environment.getId());
```

Comment:
FIXME ?

Reply:
This FIXME was already existing, and will be fixed with https://github.com/gravitee-io/issues/issues/6071